### PR TITLE
Adding a naive implementation of various primitive tensor operations

### PR DESCRIPTION
**Pull request by tannergooding**

### CODE REVIEW

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:31]: I assume the plan is to get the functionality merged and to then follow-up with vectorization in time for 8.0 as well?

Let me know if you want to divvy that up at all or if you're going to tackle it all yourself.  Presumably we'll end up wanting a `Vector<T>` implementation for downlevel and a `Vector128/256/512<T>` for .NET 8+.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:467]: This looks like it's only a subset of the approved surface. Is the plan to add the rest as part of this PR or as a follow-up?

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (tannergooding) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:31]: Yes. The general plan is to get this first bit (covering element-wise operations) in.

Work on vectorizing them can then start while the vector-based tensor operations (Signmoid, Product, Sum, etc) get their naive implementations in.

@michaelgsharp and I will be splitting the work, but you're more than welcome to join in and help as well!

Comment (tannergooding) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:31]: Getting the naive implementation in gives us both a scalar fallback and a baseline against which correctness of the SIMD implementation can be validated.

I should have tests up here in just a little bit.

Comment (stephentoub) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:31]: > @michaelgsharp and I will be splitting the work, but you're more than welcome to join in and help as well!

Happy to help

**Comment by ViktorHofer** [comment]
If the intent is to publicly ship this library going forward, you want to remove the properties in `C:\git\runtime4\src\libraries\System.Numerics.Tensors\Directory.Build.props` (and delete the file).

You will also want to update the README.md file which mentions that the library doesn't ship publicly.

**Review by michaelgsharp — APPROVED** [review]
Looks good to me. Probably still want another opinion though.

File: src/libraries/System.Numerics.Tensors/ref/System.Numerics.Tensors.csproj
Comment (ViktorHofer) [inline src/libraries/System.Numerics.Tensors/ref/System.Numerics.Tensors.csproj:4]: [Now that `NetCoreAppCurrent` means net9.0](https://github.com/dotnet/runtime/commit/85fa924eab86aedd55dc751c8ef56cf2cc66d31), we should add a net8.0 minimum TFM as well.

```suggestion
    <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```

File: src/libraries/System.Numerics.Tensors/src/System.Numerics.Tensors.csproj
Comment (ViktorHofer) [inline src/libraries/System.Numerics.Tensors/src/System.Numerics.Tensors.csproj:4]:
```suggestion
<TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:14]: Do we have any tests for boundary conditions, e.g. empty? Or error conditions?  Or is the plan for those to come separately?

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs
Comment (stephentoub) [inline src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:52]: Nit: var => float[]

**Review by stephentoub — APPROVED** [review]

**Comment by ViktorHofer** [comment]
Please respond to my above question before merging: https://github.com/dotnet/runtime/pull/91228#issuecomment-1699705294

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs
Comment (tannergooding) [reply src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:14]: I added tests for the exception cases already. I can add ones for empty as well.

For floating-point correctness cases (i.e. `infinity`, `nan`, `-0`, etc), I think it'd be good to add handling for after we move off the "naive" implementation.

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs
Comment (tannergooding) [reply src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:52]: Fixed all usages of `var`

**Comment by tannergooding** [comment]
> Please respond to my above question before merging: https://github.com/dotnet/runtime/pull/91228#issuecomment-1699705294

Removed the file that marked it as non-shipping and adjusted the readme.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs
Comment (tannergooding) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:467]: Answered in the comment thread just above.

File: src/libraries/System.Numerics.Tensors/README.md
Comment (ViktorHofer) [inline src/libraries/System.Numerics.Tensors/README.md:3]: Would be good to update this README to the format that other libraries use, i.e. https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Abstractions/README.md. This can happen in a follow-up but please don't forget about it.

Also, as this package is now shipping officially, we should add a PACKAGE readme as well. I will add an entry in https://github.com/dotnet/runtime/pull/91210 to track that work.

**Review by ViktorHofer — APPROVED** [review]

**Pull request merged**
